src/Features/Core/Portable/Completion/Providers/ImportCompletionProvider/AbstractTypeImportCompletionService.cs

```diff
@@ -51,200 +51,196 @@ public Task WarmUpCacheAsync(Project? project, CancellationToken cancellationTok
+                }
+            }
 
-            foreach (var referencedProject in referencedProjects.Where(p => p.SupportsCompilation))
+            var originCompilation = await currentProject.GetRequiredCompilationAsync(cancellationToken).ConfigureAwait(false);
+            var editorBrowsableInfo = new Lazy<EditorBrowsableInfo>(() => new EditorBrowsableInfo(originCompilation));
+            foreach (var peReference in currentProject.MetadataReferences.OfType<PortableExecutableReference>())
             {
-                var compilation = await referencedProject.GetRequiredCompilationAsync(cancellationToken).ConfigureAwait(false);
-                var assembly = SymbolFinder.FindSimilarSymbols(compilation.Assembly, currentCompilation, cancellationToken).SingleOrDefault();
-                var metadataReference = assembly != null ? currentCompilation.GetMetadataReference(assembly) : null;
+                // Can't cache items for reference with null key. We don't want risk potential perf regression by 
+                // making those items repeatedly, so simply not returning anything from this assembly, until 
+                // we have a better understanding on this scenario.
+                var key = GetPEReferenceCacheKey(peReference);
+                if (key is null || !HasGlobalAlias(peReference.Properties.Aliases))
+                    continue;
 
-                if (HasGlobalAlias(metadataReference))
+                if (CacheService.PEItemsCache.TryGetValue(key, out var cacheEntry))
                 {
-                    cacheResult = await GetCacheForProjectAsync(
-                        referencedProject,
-                        forceCacheCreation,
-                        editorBrowsableInfo: null,
-                        cancellationToken).ConfigureAwait(false);
+                    resultBuilder.Add(cacheEntry);
+                    GetUpToDateCacheForPEReferenceInBackground(currentProject.Id, peReference);
+                }
+                else if (!forceCacheCreation)
+                {
+                    isPartialResult = true;
+                    GetUpToDateCacheForPEReferenceInBackground(currentProject.Id, peReference);
+                }
+                else if (TryGetUpToDateCacheForPEReference(originCompilation, solution, editorBrowsableInfo.Value, peReference, cancellationToken, out var upToDateCacheEntry))
+                {
+                    resultBuilder.Add(upToDateCacheEntry);
+                }
+            }
 
-                    if (cacheResult.HasValue)
-                    {
-                        builder.Add(cacheResult.Value);
-                    }
-                    else
+            return (resultBuilder.ToImmutable(), isPartialResult);
+
+            void GetUpToDateCacheForProjectInBackground(ProjectId projectId)
+            {
+                lock (s_gate)
+                {
+                    s_cachingTask = s_cachingTask.ContinueWith(async _ =>
```

Comment:
i'm really not a fan of just continually creating a longer and loner chain, each of which has to redo work.  can we switch to an async batching work queue?